investos/portfolio/risk_model/factor_risk.py

```diff
@@ -1,23 +1,351 @@
+        self._recalc_freq = recalc_freq
+
+        self._fos_risk_factor_uuids = fos_risk_factor_uuids
+        self._fos_return_uuid = fos_return_uuid
+
+        self._end_date = end_date
+        self._start_date = start_date
+        self._adjustments = adjustments
+
+        if self._fos_risk_factor_uuids:
+            print("Running Barra Risk model...")
+            self._pull_fos_risk_factor_dfs()
+            self._generate_risk_models()
+            self._apply_risk_model_adjustments()
+            print(
+                "\nDone generating factor loadings,factor covariance, and idiosyncratic variance"
+            )
+        else:
+            self._factor_covariance = factor_covariance
+            self._factor_loadings = factor_loadings
+            self._idiosyncratic_variance = idiosyncratic_variance
+
+    # Initialization
+    def _pull_fos_risk_factor_dfs(self):
+        # pull from feature hub
+        print("\n1. Pulling out data from FeatureHub...")
+
+        if not fos.api_key:
+            raise ValueError(
+                "Feature hub connection failed. Please check the API key is valid."
+            )
+
+        # Initialize a DataFrame
+        self._df_loadings = (
+            fos.Feature.get(fos_return_uuid_dict[self._fos_return_uuid])
+            .get_df()
+            .rename(columns={"value": self._fos_return_uuid})
+        )
+
+        for factor in self._fos_risk_factor_uuids:
+            temp_df = (
+                fos.Feature.get(fos_risk_factor_uuids_dict[factor])
+                .get_df()
+                .rename(columns={"value": factor})
+            )
+            if "datetime" not in temp_df.columns:
+                # Perform operations for non-datetime columns (assuming to create dummies)
+                temp_df = pd.get_dummies(temp_df, columns=[factor], dtype=int)
+                self._df_loadings = self._df_loadings.merge(
+                    temp_df, how="left", on="id"
+                )
+                self._fos_risk_factor_uuids.remove(factor)
+            else:
+                self._df_loadings = pd.merge_asof(
+                    self._df_loadings.sort_values("datetime"),
+                    temp_df.sort_values("datetime"),
+                    on="datetime",
+                    by="id",
+                    direction="forward",
+                )
+        # handling missing data and inf data
+        self._df_loadings = self._df_loadings.replace(
+            [np.inf, -np.inf], np.nan
+        ).dropna()
+        self._df_loadings = self._df_loadings[
+            (self._df_loadings.datetime >= self._end_date)
+            & (self._df_loadings.datetime <= self._start_date)
+        ]
+
+        print("Done pulling data")
+        return
+
+    def _generate_risk_models(self, timestamp="datetime") -> None:
+        """fit a cross sectional factor model and generate factor return covariance and idio returns variance
 
-        self._drop_excluded_assets()
+        Parameters
```

Comment:
These parameters are wrong. They look like they've been copy pasted from somewhere else.

Reply:
(The docstring for the param description is wrong is specifically what I mean)